tools/ci/flink-ci-tools/src/test/java/org/apache/flink/tools/ci/optional/ShadeOptionalCheckerTest.java

```diff
@@ -0,0 +1,157 @@
+    }
+
+    @Test
+    void testDirectBundledDependencyMustBeOptional() {
+        final Dependency dependency = createMandatoryDependency("a");
+        final Set<Dependency> bundled = Collections.singleton(dependency);
+        final DependencyTree dependencyTree = new DependencyTree().addDirectDependency(dependency);
+
+        final Set<Dependency> violations =
+                ShadeOptionalChecker.checkOptionalFlags(MODULE, bundled, dependencyTree);
+
+        assertThat(violations).containsExactly(dependency);
+    }
+
+    @Test
+    void testTransitiveBundledOptionalDependencyIsAccepted() {
+        final Dependency dependencyA = createMandatoryDependency("a");
+        final Dependency dependencyB = createOptionalDependency("b");
+        final Set<Dependency> bundled = Collections.singleton(dependencyB);
+        final DependencyTree dependencyTree =
+                new DependencyTree()
+                        .addDirectDependency(dependencyA)
+                        .addTransitiveDependencyTo(dependencyB, dependencyA);
+
+        final Set<Dependency> violations =
+                ShadeOptionalChecker.checkOptionalFlags(MODULE, bundled, dependencyTree);
+
+        assertThat(violations).isEmpty();
+    }
+
+    @Test
+    void testTransitiveBundledDependencyMustBeOptional() {
+        final Dependency dependencyA = createMandatoryDependency("a");
+        final Dependency dependencyB = createMandatoryDependency("b");
+        final Set<Dependency> bundled = Collections.singleton(dependencyB);
+        final DependencyTree dependencyTree =
+                new DependencyTree()
+                        .addDirectDependency(dependencyA)
+                        .addTransitiveDependencyTo(dependencyB, dependencyA);
+
+        final Set<Dependency> violations =
+                ShadeOptionalChecker.checkOptionalFlags(MODULE, bundled, dependencyTree);
+
+        assertThat(violations).containsExactly(dependencyB);
+    }
+
+    @Test
+    void testTransitiveBundledDependencyMayNotBeOptionalIfParentIsOptional() {
+        final Dependency dependencyA = createOptionalDependency("a");
+        final Dependency dependencyB = createMandatoryDependency("b");
+        final Set<Dependency> bundled = Collections.singleton(dependencyB);
+        final DependencyTree dependencyTree =
+                new DependencyTree()
+                        .addDirectDependency(dependencyA)
+                        .addTransitiveDependencyTo(dependencyB, dependencyA);
+
+        final Set<Dependency> violations =
+                ShadeOptionalChecker.checkOptionalFlags(MODULE, bundled, dependencyTree);
+
+        assertThat(violations).isEmpty();
+    }
+
+    @Test
+    void testTestDependenciesAreIgnored() {
+        final Dependency dependencyA = createTestDependency("a");
+        final Set<Dependency> bundled = Collections.emptySet();
+        final DependencyTree dependencyTree = new DependencyTree().addDirectDependency(dependencyA);
```

Comment:
This test still doesn't check anything. `dependencyA` could be a mandatory or optional dependency and the test would still succeed. Should we check that a mandatory transitive dependency is ignored even if it is supposed to be bundled as described in the following proposal:
```suggestion
        final Dependency dependencyB = createMandatoryDependency("b");
        final Set<Dependency> bundled = Collections.singleton(dependencyB);
        final DependencyTree dependencyTree = new DependencyTree().addDirectDependency(dependencyA).addTransitiveDependencyTo(dependencyB, dependencyA);
```